documents/ros.1
.TH "ros" "1" "" "" ""
.nh \" Turn off hyphenation by default.
.PP
Roswell \- Common Lisp environment setup Utility
.SH Synopsis
.IP \[bu] 2
\f[B]ros\f[] [\f[B]options\f[]] \f[C][command\ [args...]\f[]
.IP \[bu] 2
\f[B]ros\f[] [\f[B]options\f[]] \f[C][\-\-]script\ [args...]\f[]
.IP \[bu] 2
\f[B]ros\f[] [\f[B]options\f[]] < script.ros
.SH Description
.PP
Roswell is a command line tool for installing and managing Common Lisp
implementations, as well as a scripting infrastructure for templating,
writing, executing, compiling, distributing, downloading and installing
ROSWELL\-SCRIPTS written in Common Lisp.
.PP
Currently roswell supports sbcl, ccl, clisp and ecl as its supported
lisp implementations.
For further details see \f[I]ros\-install(1)\f[].
.PP
Scripts installed by roswell will be system wide if appropriate
directory is included in PATH.
.SH Subcommands
.PP
In an order of utility/frequency.
.TP
.B install
Install a given implementation (e.g.
sbcl, ccl) or a system (e.g.
alexandria) for roswell environment.
See \f[I]ros\-install(1)\f[].
.RS
.RE
.TP
.B init [name[.ros]]
Create a new ros script
.RS
.RE
.TP
.B dump [executable|output] [script]
Dump an image of the script for the faster startup or to make an
executable.
.RS
.RE
.TP
.B build
Make an executable from the script.
.RS
.RE
.TP
.B run
Initiate REPL
.RS
.RE
.TP
.B use
Change the default implementation used by roswell.
.RS
.RE
.TP
.B list
List the various informations.
.RS
.RE
.TP
.B config
Get and set the options.
.RS
.RE
.TP
.B setup
Run the initial setup
.RS
.RE
.TP
.B emacs
Launch emacs with slime.
.RS
.RE
.TP
.B wait
Wait forever, used for daemonizing the script.
.RS
.RE
.TP
.B delete
Delete an installed implementation.
.RS
.RE
.TP
.B version
Show the roswell version information.
.RS
.RE
.TP
.B help
Show the subcommand help.
.RS
.RE
.SH Options
.PP
Options are processed in left\-to\-right order.
.TP
.B \-w CODE \-\-wrap CODE
Run the script with a shell wrapper CODE, e.g.
rlwrap
.RS
.RE
.TP
.B \-m IMAGE \-\-image IMAGE
Continue from Lisp image IMAGE
.RS
.RE
.TP
.B \-L NAME \-\-lisp NAME
Run the script with a lisp impl NAME[/VERSION] if present, e.g.
sbcl\-bin, sbcl/1.2.16.
Fails otherwise.
.RS
.RE
.TP
.B \-l FILE \-\-load FILE
Load a lisp file FILE while building
.RS
.RE
.TP
.B \-S X \-\-source\-registry X
Override the source registry of asdf systems.
.RS
.RE
.TP
.B \-s SYSTEM \-\-system SYSTEM
Load the asdf SYSTEM while building.
.RS
.RE
.TP
.B \-\-load\-system SYSTEM
Same as above (buildapp compatibility)
.RS
.RE
.TP
.B \-p PACKAGE \-\-package PACKAGE
Change the current package to PACKAGE
.RS
.RE
.TP
.B \-sp SP \-\-system\-package SP
Combination of \-s SP and \-p SP
.RS
.RE
.TP
.B \-e FORM \-\-eval FORM
Evaluate FORM while building
.RS
.RE
.TP
.B \-\-require MODULE
require MODULE while building
.RS
.RE
.TP
.B \-q \-\-quit
quit lisp here
.RS
.RE
.TP
.B \-r FUNC \-\-restart FUNC
the build image restarts from calling FUNC
.RS
.RE
.TP
.B \-E FUNC \-\-entry FUNC
the build image restarts from calling (FUNC argv).
.RS
.RE
.TP
.B \-i FORM \-\-init FORM
evaluate FORM after the restart.
.RS
.RE
.TP
.B \-ip FORM \-\-print FORM
evaluate and princ FORM after the restart
.RS
.RE
.TP
.B \-iw FORM \-\-write FORM
evaluate and write FORM after the restart
.RS
.RE
.TP
.B \-F FORM \-\-final FORM
evaluate FORM before dumping the IMAGE
.RS
.RE
.TP
.B \-R \-\-rc
try to read /etc/rosrc, ~/.roswell/init.lisp
.RS
.RE
.TP
.B +R \-\-no\-rc
skip /etc/rosrc, ~/.roswell/init.lisp
.RS
.RE
.TP
.B \-Q \-\-quicklisp
load quicklisp (default)
.RS
.RE
.TP
.B +Q \-\-no\-quicklisp
do not load quicklisp
.RS
.RE
.TP
.B \-v \-\-verbose
be quite verbose while building
.RS
.RE
.TP
.B \-\-quiet
suppress output while building (default)
.RS
.RE
.TP
.B \-\-test
for test purpose
.RS
.RE
.TP
.B dynamic\-space\-size=[size in MB]
SBCL specific.
The argument is passed to SBCL by \f[C]\-\-dynamic\-space\-size\f[]
.RS
.RE
.SH Environmental Variables
.TP
.B ROSWELL_INSTALL_DIR
Specifies the install directory of roswell, defaulted to $HOME/.roswell
.
Roswell scripts are installed in $ROSWELL_INSTALL_DIR/bin .
.RS
.RE
.TP
.B ROSWELL_BRANCH
used for testing roswell itself.
.RS
.RE
.TP
.B ROSWELL_REPO
used for testing roswell itself.
.RS
.RE
.TP
.B LISP_IMPLS_DIR
used for testing roswell itself.
.RS
.RE
.TP
.B LISP_IMPLS_BIN
used for testing roswell itself.
.RS
.RE
.TP
.B ROSWELL_TARBALL_PATH
used for testing roswell itself.
.RS
.RE
.TP
.B ROSWELL_URI
used for testing roswell itself.
.RS
.RE
.SH Bugs
.PP
Check out issues list (https://github.com/snmsts/roswell/issues)
.SH SEE ALSO
.PP
\f[I]sbcl\f[](1) \f[I]ros\-dump\f[](1) \f[I]ros\-init\f[](1)
\f[I]ros\-install\f[](1) \f[I]ros\-list\f[](1) \f[I]ros\-setup\f[](1)
.SH AUTHORS
Masataro Asai.


ChangeLog
roswell (0.0.3.54-1) unstable; urgency=low

  * New: `ros:ensure-using-downloaded-asdf` are introduced.
  * New: Use the latest ASDF for CI when $ROSWELL_LATEST_ASDF is true.
  * Change: Use pandoc to to generate documents man and html.
  * Improve: Better support for armv7l.
  * Improve: More documentation.(by Masataro Asai)

 -- SANO Masatoshi <snmsts@gmail.com>  Sat, 19 Dec 2015 20:55:22 +0900

roswell (0.0.3.53-1) unstable; urgency=low

  * New: support installing ecl.
  * Improve: show error message when download page are not include 'sbcl'.
  * Change: 'ros init' make a file with package.

 -- SANO Masatoshi <snmsts@gmail.com>  Thu, 26 Nov 2015 13:53:45 +0900

roswell (0.0.3.52-1) unstable; urgency=low

  * Improve: try support arm64.(just uname-> architecture)
  * Improve: use msys2 are alredy running, use it.
  * Improve: write some 'man' files.
  * Fix: 'ros build script' now don't take default option for sbcl. (reported by noctuid)

 -- SANO Masatoshi <snmsts@gmail.com>  Sat, 14 Nov 2015 19:02:06 +0900

roswell (0.0.3.51-1) unstable; urgency=low

  * Improve: hack compile-file*.
  * Fix: pararel compile groff files.

 -- SANO Masatoshi <snmsts@gmail.com>  Tue, 03 Nov 2015 23:05:53 +0900

roswell (0.0.3.50-1) unstable; urgency=low

  * New: New subcommand wait are implemented. 'ros wait' will good to use with server application.
  * Fix: install-for-ci.sh update link to acl version to 10.0.
  * Fix: armhf platform fixed.(by Pavel)
  * Improve: re-install quicklisp init file if roswell updated.
  * Improve: hack quicklisp to use roswell to extract tar.gz archive.
  * Improve: messages and documentations are written and reviewed. (by Masataro Asai)

 -- SANO Masatoshi <snmsts@gmail.com>  Fri, 30 Oct 2015 11:04:08 +0900

roswell (0.0.3.49-1) unstable; urgency=low

  * New: Now 'ros install sbcl' work on windows.
  * Fix: fix determin architecture broken for armv7l reported by pimiento.
  * Improve: start working install-for-ci.PS1.
  * Improve: make all embeded URI configurable.

 -- SANO Masatoshi <snmsts@gmail.com>  Thu, 08 Oct 2015 02:48:15 +0900

roswell (0.0.3.48-1) unstable; urgency=low

  * New: 'ros list versions impl' would show a version list installeable.
  * Improve: check atom file for checking sbcl version.

 -- SANO Masatoshi <snmsts@gmail.com>  Sun, 20 Sep 2015 22:39:42 +0900

roswell (0.0.3.47-1) unstable; urgency=low

  * Fixed: Wrong implementation's dump  are listed when 'ros list dump'.
  * Improve: 'ros help' can show more subcommands.
  * Fixed: Ensure 'ros run' invoke 'ros setup' if not yet setup.then run sbcl-bin.
  * Improve: Stop manual generation by default.
  * Improve: configure script takes '--with-sbcl-bin-base=URI' to specify sbcl-bin's base uri (by Masaya TANIGUCHI)

 -- SANO Masatoshi <snmsts@gmail.com>  Mon, 14 Sep 2015 01:02:43 +0900

roswell (0.0.3.46-1) unstable; urgency=low

  * Fixed: support ASDF 3.1.5 change asdf::*user-cache*

 -- SANO Masatoshi <snmsts@gmail.com>  Tue, 01 Sep 2015 19:01:28 +0900

roswell (0.0.3.45-1) unstable; urgency=low

  * Fixed: better algorithm to calc download progress.
  * Fixed: Cache directory conflicts when sbcl and sbcl-bin are just same version.
  * Fixed: wait for input forever when try to compile sbcl again on Darwin.
  * Improve: Better support for CI service using Darwin.

 -- SANO Masatoshi <snmsts@gmail.com>  Sun, 30 Aug 2015 14:29:54 +0900

roswell (0.0.3.44-1) unstable; urgency=low

  * Fixed:Error when installing ros script from locally installed project.
  * Improve: scripts/install-for-ci.sh improved especially ABCL and CMUCL.(by Eitaro Fukamachi)
  * New: on windows, 'ros install msys2' implemented.

 -- SANO Masatoshi <snmsts@gmail.com>  Sat, 22 Aug 2015 14:23:18 +0900

roswell (0.0.3.43-1) unstable; urgency=low

  * Improve: Confirmed and some fix are applied for scripts/install-for-ci.sh

 -- SANO Masatoshi <snmsts@gmail.com>  Sat, 15 Aug 2015 18:59:27 +0900

roswell (0.0.3.42-1) unstable; urgency=low

  * New: support new impls for 'ros run' (clisp ecl abcl cmu acl)
  * Improve: Better error handling in 'ros build' (by Eitaro Fukamachi)
  * Improve: scripts/install-for-ci.sh improved for new impls (by Eitaro Fukamachi)
  * Change: 'ros install package' will call function bound to 'ros.install:*build-hook*'

 -- SANO Masatoshi <snmsts@gmail.com>  Thu, 13 Aug 2015 23:43:32 +0900

roswell (0.0.3.41-1) unstable; urgency=low

  * Change: use github to download sbcl source.

 -- SANO Masatoshi <snmsts@gmail.com>  Fri, 31 Jul 2015 09:05:16 +0900

roswell (0.0.3.40-1) unstable; urgency=low

  * New: support download failure cause from upstream upload failure(?)

 -- SANO Masatoshi <snmsts@gmail.com>  Thu, 30 Jul 2015 16:50:18 +0900

roswell (0.0.3.39-1) unstable; urgency=low

  * Change: revert automatically compile ros script facility (0.0.3.38-1).
  * New: Handle SIGINT
  * New: 'ros install sbcl/ccl-bin --without-install' will just download archive.

 -- SANO Masatoshi <snmsts@gmail.com>  Tue, 28 Jul 2015 03:54:55 +0900

roswell (0.0.3.38-1) unstable; urgency=low

  * Change: add gist url for outage of sbcl.org.
  * New: scripts/make-table.ros to make apropriate html file for roswell.
  * Change: ros script are automatically compiled when it is installed by 'ros install hoge'.
  * Improve: it's experimental implementation but setuid/setgid are properly done for 'ros setup'.

 -- SANO Masatoshi <snmsts@gmail.com>  Thu, 23 Jul 2015 00:54:46 +0900

roswell (0.0.3.37-1) unstable; urgency=low

  * Change: Change progress bar width (for CMD.exe WIN32 platform)
  * Bugfix: Fix typo of --load-system option (by Eitaro Fukamachi)
  * Bugfix: 'ros -L ccl-bin run' fails when environment variable
	'CCL_DEFAULT_DIRECTORY' are set (reported by Tatsuya BIZENN)
  * Bugfix: fix some memory leaky part.

 -- SANO Masatoshi <snmsts@gmail.com>  Sun, 28 Jun 2015 10:32:04 +0900

roswell (0.0.3.36-1) unstable; urgency=low

  * Bugfix: Build fail without git.


src/ros.c
#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif
#include "opt.h"

char** argv_orig;
int argc_orig;
struct opts* global_opt;
struct opts* local_opt=NULL;

extern int cmd_list(int argc,char **argv,struct sub_command* cmd);
extern int cmd_version(int argc,char **argv,struct sub_command* cmd);
extern int cmd_config(int argc,char **argv,struct sub_command* cmd);
extern int cmd_setup(int argc,char **argv,struct sub_command* cmd);
extern int cmd_help(int argc,char **argv,struct sub_command* cmd);
extern int cmd_internal(int argc,char **argv,struct sub_command* cmd);

extern void register_cmd_run(void);
extern void register_cmd_install(void);
extern void register_cmd_config(void);
extern void register_cmd_setup(void);
extern void register_cmd_internal(void);

int verbose=0;
int testing=0;
int rc=1;
int quicklisp=1;

LVal top_commands =(LVal)NULL;
LVal top_options =(LVal)NULL;

LVal top_helps =(LVal)NULL;
LVal subcommand_name=(LVal)NULL;

int proccmd(int argc,char** argv,LVal option,LVal command);

int proccmd_with_subcmd(char* path,char* subcmd,int argc,char** argv,LVal option,LVal command) {
  char** argv2=(char**)alloc(sizeof(char*)*(argc+2));
  int i,ret;
  for(i=0;i<argc;++i) {
    argv2[i+2]=argv[i];
  }
  argv2[0]=path;
  argv2[1]=subcmd;
  ret=proccmd(argc+2,argv2,option,command);
  dealloc(argv2);
  return ret;
}

int proccmd(int argc,char** argv,LVal option,LVal command) {
  int pos;
  cond_printf(1,"proccmd:%s\n",argv[0]);
  if(argv[0][0]=='-' || argv[0][0]=='+') {
    if(argv[0][0]=='-' &&
       argv[0][1]=='-') { /*long option*/
      LVal p;
      for(p=option;p;p=Next(p)) {
        struct sub_command* fp=firstp(p);
        if(strcmp(&argv[0][2],fp->name)==0) {
          int result= fp->call(argc,argv,fp);
          if(fp->terminating) {
            cond_printf(1,"terminating:%s\n",argv[0]);
            exit(result);
          }else
            return result;
        }
      }
      cond_printf(1,"proccmd:invalid? %s\n",argv[0]);
    }else { /*short option*/
      if(argv[0][1]!='\0') {
        LVal p;
        for(p=option;p;p=Next(p)) {
          struct sub_command* fp=firstp(p);
          if(fp->short_name&&strcmp(argv[0],fp->short_name)==0) {
            int result= fp->call(argc,argv,fp);
            if(fp->terminating) {
              cond_printf(1,"terminating:%s\n",argv[0]);
              exit(result);
            }else {
              return result;
            }
          }
        }
      }
      /* invalid */
    }
  }else if((pos=position_char("=",argv[0]))!=-1) {
    char *l,*r;
    l=subseq(argv[0],0,pos);
    r=subseq(argv[0],pos+1,0);
    if(r)
      set_opt(&local_opt,l,r,0);
    else{
      struct opts* opt=global_opt;
      struct opts** opts=&opt;
      unset_opt(opts, l);
      global_opt=*opts;
    }
    s(l),s(r);
    return proccmd(argc-1,&(argv[1]),option,command);
  }else {
    char* tmp[]={"help"};
    LVal p;
    if(command==top_commands && position_char(".",argv[0])==-1) {
      /* local commands*/
      char* cmddir=configdir();
      char* cmdpath=cat(cmddir,argv[0],".ros",NULL);
      if(directory_exist_p(cmddir) && file_exist_p(cmdpath))
        proccmd_with_subcmd(cmdpath,"main",argc,argv,top_options,top_commands);
      s(cmddir),s(cmdpath);
      /* systemwide commands*/
      cmddir=subcmddir();
      cmdpath=cat(cmddir,argv[0],".ros",NULL);
      if(directory_exist_p(cmddir)) {
        if(file_exist_p(cmdpath))
          proccmd_with_subcmd(cmdpath,"main",argc,argv,top_options,top_commands);
        s(cmdpath);cmdpath=cat(cmddir,"+",argv[0],".ros",NULL);
        if(file_exist_p(cmdpath))
          proccmd_with_subcmd(cmdpath,"main",argc,argv,top_options,top_commands);
      }
      s(cmddir),s(cmdpath);
    }
    /* search internal commands.*/
    for(p=command;p;p=Next(p)) {
      struct sub_command* fp=firstp(p);
      if(fp->name&&(strcmp(fp->name,argv[0])==0||strcmp(fp->name,"*")==0)) {
        subcommand_name=conss((char*)fp->name,subcommand_name);
        exit(fp->call(argc,argv,fp));
      }
    }
    fprintf(stderr,"invalid command\n");
    proccmd(1,tmp,top_options,top_commands);
  }
  return 1;
}

int opt_top(int argc,char** argv,struct sub_command* cmd) {
  const char* arg=cmd->name;
  if(arg && argc>1)
    set_opt(&local_opt,arg,argv[1],0);
  return 2;
}

int opt_top_verbose(int argc,char** argv,struct sub_command* cmd) {
  if(strcmp(cmd->name,"verbose")==0) {
    verbose=1|verbose<<1;
  }else if(strcmp(cmd->name,"quiet")==0) {
    verbose=verbose>>1;
  }
  cond_printf(1,"opt_verbose:%s %d\n",cmd->name,verbose);
  return 1;
}

int opt_top_testing(int argc,char** argv,struct sub_command* cmd) {
  ++testing;
  cond_printf(1,"opt_testing:%s %d\n",cmd->name,testing);
  return 1;
}

int opt_top_rc(int argc,char** argv,struct sub_command* cmd) {
  if(strcmp(cmd->name,"rc")==0) {
    rc=1;
  }else if(strcmp(cmd->name,"no-rc")==0)
    rc=0;
  cond_printf(1,"opt_rc:%s\n",cmd->name);
  return 1;
}

int opt_top_ql(int argc,char** argv,struct sub_command* cmd) {
  if(strcmp(cmd->name,"quicklisp")==0) {
    quicklisp=1;
  }else if(strcmp(cmd->name,"no-quicklisp")==0)
    quicklisp=0;
  cond_printf(1,"opt_quicklisp:%s\n",cmd->name);
  return 1;
}

int opt_top_build0(int argc,char** argv,struct sub_command* cmd) {
  if(cmd->name) {
    char* current=get_opt("program",0);
    current=cat(current?current:"","(:",cmd->name,")",NULL);
    set_opt(&local_opt,"program",current,0);
  }
  return 1;
}

int opt_top_build(int argc,char** argv,struct sub_command* cmd) {
  if(cmd->name && argc>1) {
    char* current=get_opt("program",0);
    char* escaped=escape_string(argv[1]);
    current=cat(current?current:"","(:",cmd->name," \"",escaped,"\")",NULL);
    s(escaped);
    set_opt(&local_opt,"program",current,0);
  }
  return 2;
}

int opt_restart_after(int argc,char** argv,struct sub_command* cmd) {
  if(cmd->name && argc>1) {
    char* current=get_opt("restart",0);
    char* escaped=escape_string(argv[1]);
    current=cat(current?current:"","(:",cmd->name," \"",escaped,"\")",NULL);
    s(escaped);
    set_opt(&local_opt,"restart",current,0);
  }
  return 2;
}

int opt_final(int argc,char** argv,struct sub_command* cmd) {
  if(cmd->name && argc>1) {
    char* current=get_opt("final",0);
    char* escaped=escape_string(argv[1]);
    current=cat(current?current:"","(:",cmd->name," \"",escaped,"\")",NULL);
    s(escaped);
    set_opt(&local_opt,"final",current,0);
  }
  return 2;
}

LVal register_runtime_options(LVal opt) {
  /*opt=add_command(opt,"file","-f",opt_top_build,1,0,"include lisp FILE while building","FILE");*/
  opt=add_command(opt,"load","-l",opt_top_build,1,0,"load lisp FILE while building","FILE");
  opt=add_command(opt,"source-registry","-S",opt_top_build,1,0,"override source registry of asdf systems","X");
  opt=add_command(opt,"system","-s",opt_top_build,1,0,"load asdf SYSTEM while building","SYSTEM");
  opt=add_command(opt,"load-system",NULL,opt_top_build,1,0,"same as above (buildapp compatibility)","SYSTEM");
  opt=add_command(opt,"package","-p",opt_top_build,1,0,"change current package to PACKAGE","PACKAGE");
  opt=add_command(opt,"system-package","-sp",opt_top_build,1,0,"combination of -s SP and -p SP","SP");
  opt=add_command(opt,"eval","-e",opt_top_build,1,0,"evaluate FORM while building","FORM");
  opt=add_command(opt,"require",NULL,opt_top_build,1,0,"require MODULE while building","MODULE");
  opt=add_command(opt,"quit","-q",opt_top_build0,1,0,"quit lisp here",NULL);

  opt=add_command(opt,"restart","-r",opt_restart_after,1,0,"restart from build by calling (FUNC)","FUNC");
  opt=add_command(opt,"entry","-E",opt_restart_after,1,0,"restart from build by calling (FUNC argv)","FUNC");
  opt=add_command(opt,"init","-i",opt_restart_after,1,0,"evaluate FORM after restart","FORM");
  opt=add_command(opt,"print","-ip",opt_restart_after,1,0,"evaluate and princ FORM after restart","FORM");
  opt=add_command(opt,"write","-iw",opt_restart_after,1,0,"evaluate and write FORM after restart","FORM");

  opt=add_command(opt,"final","-F",opt_final,1,0,"evaluate FORM before dumping IMAGE","FORM");

  /* opt=add_command(opt,"include","-I",cmd_notyet,1,0,"runtime PATH to cl-launch installation","PATH"); */
  /* opt=add_command(opt,"no-include","+I",cmd_notyet,1,0,"disable cl-launch installation feature",NULL); */
  opt=add_command(opt,"rc","-R",opt_top_rc,1,0,"try read /etc/rosrc, ~/.roswell/init.lisp",NULL);
  opt=add_command(opt,"no-rc","+R",opt_top_rc,1,0,"skip /etc/rosrc, ~/.roswell/init.lisp",NULL);
  opt=add_command(opt,"quicklisp","-Q",opt_top_ql,1,0,"use quicklisp (default)",NULL);
  opt=add_command(opt,"no-quicklisp","+Q",opt_top_ql,1,0,"do not use quicklisp",NULL);
  opt=add_command(opt,"verbose","-v",opt_top_verbose,1,0,"be quite noisy while building",NULL);
  opt=add_command(opt,"quiet",NULL,opt_top_verbose,1,0,"be quite quiet while building (default)",NULL);
  opt=add_command(opt,"test",NULL,opt_top_testing,1,0,"for test purpose",NULL);
  opt=add_command(opt,"stdin",NULL,opt_top_build,0,0,NULL,NULL);
  return opt;
}

int main (int argc,char **argv) {
  argv_orig=argv;
  argc_orig=argc;
  char* _help;
  /*options*/
  /* toplevel */
  top_options=add_command(top_options,"wrap","-w",opt_top,1,0,"Run "PACKAGE" with a shell wrapper CODE,","CODE");
  top_options=add_command(top_options,"image","-m",opt_top,1,0,"continue from Lisp image IMAGE","IMAGE");
  top_options=add_command(top_options,"lisp","-L",opt_top,1,0,"Run "PACKAGE" with a lisp impl NAME[/VERSION].","NAME");
  top_options=register_runtime_options(top_options);

  /* abbrevs */
  top_options=add_command(top_options,"version","-V",cmd_version,0,1,NULL,NULL);
  top_options=add_command(top_options,"help","-h",cmd_help,0,1,NULL,NULL);
  top_options=add_command(top_options,"help","-?",cmd_help,0,1,NULL,NULL);

  top_options=nreverse(top_options);
  /*commands*/
  register_cmd_install();
  top_commands=add_command(top_commands,"roswell-internal-use",NULL,cmd_internal,0,1,NULL,NULL);
  register_cmd_config();
  register_cmd_setup();

  /*         {"list",NULL,cmd_list,1,1}, */
  /*         {"set",NULL,cmd_notyet,0,1}, */
  top_commands=add_command(top_commands,"version" ,NULL,cmd_version,1,1,"Show the "PACKAGE" version information",NULL);
  register_cmd_internal();

  top_commands=add_command(top_commands,"help",NULL,cmd_help,1,1,"Show Command help",NULL);
  register_cmd_run();
  top_commands=nreverse(top_commands);
  _help=cat("Usage: ",argv_orig[0]," [OPTIONS] [Command arguments...]  \n",
            "Usage: ",argv_orig[0]," [OPTIONS] [[--] script-path arguments...]  \n\n",NULL);
  top_helps=add_help(top_helps,NULL,_help,top_commands,top_options,NULL,NULL,NULL);
  s(_help);

  char* path=s_cat(configdir(),q("config"),NULL);
  global_opt=load_opts(path);
  struct opts** opts=&global_opt;
  unset_opt(opts,"program");
  s(path);
  if(argc==1) {
    char* tmp[]={"help"};
    proccmd(1,tmp,top_options,top_commands);
  }else {
    int i;
    for(i=1;i<argc;i+=proccmd(argc-i,&argv[i],top_options,top_commands));
  }
  if(get_opt("program",0)) {
    char* tmp[]={"run","-q","--"};
    proccmd(3,tmp,top_options,top_commands);
  }
  free_opts(global_opt);
}
